File: 0068-text-justification/0068-text-justification.java
class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> res = new ArrayList<>();
        List<String> temp = new ArrayList<>();
        int lineLength = 0;

        for(String word : words)
        {
            // 检查每行的字符数是否超过最大宽度
            // temp.size() 为每行空格数量
            if(lineLength + temp.size() + word.length() > maxWidth)
            {
                res.add(justifySpaces(temp, lineLength, maxWidth));
                temp.clear();
                lineLength = 0;
            }

            temp.add(word);
            lineLength += word.length();
        }

        // Handle the last line (left-justified)
        StringBuilder lastLine = new StringBuilder(String.join(" ", temp));
        while (lastLine.length() < maxWidth) {
            lastLine.append(" ");
        }
        res.add(lastLine.toString());

        return res;

    }

    private String justifySpaces(List<String> line, int lineLength, int maxWidth)
    {
        int spacesToDistribute = maxWidth - lineLength; // 计算出还剩多少个空格可以分配
        int gaps = line.size() - 1;                     // 计算出"缺口"的数量

        // 如果这一行只有一个word
        if(gaps == 0)
        {
            StringBuilder sb = new StringBuilder(line.get(0));
            while(sb.length() < maxWidth)
            {
                sb.append(" ");
            }
            return sb.toString();
        }

        int spacePerGap = spacesToDistribute / gaps;
        int extraSpaces = spacesToDistribute % gaps;

        StringBuilder justifiedLine = new StringBuilder();

        for(int i = 0 ; i < gaps ; i++)
        {
            justifiedLine.append(line.get(i));

            // 添加空格
            for(int j = 0 ; j < spacePerGap ; j++)
            {
                justifiedLine.append(" ");
            }

            if(extraSpaces > 0)
            {
                justifiedLine.append(" ");
                extraSpaces--;
            }
        }

        justifiedLine.append(line.get(gaps));
        return justifiedLine.toString();
    }
}

File: 0068-text-justification/0068-text-justification.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> result;
        int i = 0, n = words.size();

        while(i < n){
            int lineLen = words[i].size();
            int j = i + 1; // 下一个不能放下的单词的索引

            // 找出能放下的单词
            while(j < n && lineLen + words[j].size() + (j - i) <= maxWidth){
                lineLen += words[j].size();
                j++;
            }

            int wordCount = j - i;
            int totalSpaces = maxWidth - lineLen;
            string line;

            // 最后一行 或 某一行只有一个单词 -> 左对齐
            if(j == n || wordCount == 1){
                for(int k = i ; k < j ; k++){
                    line += words[k];
                    if(k != j - 1) line += " ";
                }
                line += string(maxWidth - line.size(), ' ');
            } else {
                // 其他行平均分配空格
                int spacePerSlot = totalSpaces / (wordCount - 1);
                int extraSpaces = totalSpaces % (wordCount - 1); // 余数就是多出来的空格数 (余数肯定小于空隙数)

                for(int k = i ; k < j - 1 ; k++){
                    line += words[k];
                    line += string(spacePerSlot + (k - i < extraSpaces ? 1 : 0), ' '); // k - i 表示第几个空隙
                }
                line += words[j - 1]; // 最后一个单词不加空格
            }

            result.push_back(line);
            i = j; // 处理下一行
        }

        return result;
    }
};